src/pages/experiments/[experimentId]/data-processing/configure-embedding/components/EmbeddingPreview.jsx

```diff
@@ -40,99 +41,131 @@ import LabelsDesign from '../../../plots-and-tables/components/LabelsDesign';
     },
     mitochondrialFraction: {
       title: 'Mitochondrial fraction reads',
-      initialConfig: initialPlotConfigStates.embeddingContinuous,
       specGenerator: generateEmbeddingContinuousSpec,
       imgSrc: plot2Pic,
+      plotUuid: 'embeddingPreviewMitochondrialReads',
+      plotType: 'embeddingContinuous',
     },
     doubletScore: {
       title: 'Cell doublet score',
-      initialConfig: initialPlotConfigStates.embeddingContinuous,
       specGenerator: generateEmbeddingContinuousSpec,
       imgSrc: plot2Pic,
+      plotUuid: 'embeddingPreviewDoubletScore',
+      plotType: 'embeddingContinuous',
     },
   };
 
+  const config = useSelector((state) => state.componentConfig[plots[selectedSpec].plotUuid]?.config);
+
+  // Prepare data for categorical embedding
+  const embeddingType = 'umap';
+
+  const cellSets = useSelector((state) => state.cellSets);
+  const { data } = useSelector((state) => state.embeddings[embeddingType]) || {};
+  const selectedExpression = useSelector((state) => state.genes.expression.data[config?.shownGene]);
+
+  useEffect(() => {
+    if (!data) {
+      dispatch(loadEmbedding(experimentId, embeddingType));
+      dispatch(loadCellSets(experimentId));
+    }
+  }, [experimentId]);
+
   useEffect(() => {
     // Do not update anything if the cell sets are stil loading or if
     // the config does not exist yet.
+    if (!config || !data) {
+      return;
+    }
 
-    if (!config) {
+    if (!selectedExpression && config.shownGene) {
+      dispatch(loadGeneExpression(experimentId, [config.shownGene]));
+    }
+
+    if ((plots[selectedSpec].plotType === 'embeddingContinuous' && !selectedExpression)
+      || !cellSets) {
       return;
     }
 
     const spec = plots[selectedSpec].specGenerator(config);
-    generateData(spec);
+
+    if (plots[selectedSpec].plotType === 'embeddingContinuous') {
+      // If embedding is continuous
+
+      spec.data.forEach((s) => {
+        if (s.name === 'expression') {
+          s.values = selectedExpression;
+        } else if (s.name === 'embedding') {
+          s.values = data;
+        }
+      });
+    } else {
+      // If embedding categorical
+      // First, find the child nodes in the hirerarchy.
+      let newCellSets = cellSets.hierarchy.find(
+        (rootNode) => rootNode.key === config.selectedCellSet,
+      )?.children || [];
+
+      // Build up the data source based on the properties. Note that the child nodes
+      // in the hierarchy are /objects/ with a `key` property, hence the destructuring
+      // in the function.
+      newCellSets = newCellSets.map(({ key }) => ({
+        cellSetId: key,
+        ...cellSets.properties[key],
+        cellIds: Array.from(cellSets.properties[key].cellIds),
+      }));
+
+      spec.data.forEach((s) => {
+        if (s.name === 'cellSets') {
+          s.values = newCellSets;
+        } else if (s.name === 'embedding') {
+          s.values = data;
+        }
+      });
+    }
+
```

Comment:
better to refactor this into a separate function

Reply:
Refactored in new commit